### DISCUSSION TITLE
How to listen for application close (X) event on PC

### DISCUSSION THREAD

**Discussion by NgVThangBz** [axmolengine/axmol#2285] [Q&A] [answered]
i need to free my data before closing my app before the app is closed by (X),
so how can i listen to the app close event.
i used `glView->windowShouldClose()`, however it doesn't seem to be what i want,

and is it possible to block the close event, I want to show a message "do you want to close the app?" before the app is closed on PC

**Comment by rh101** [comment]
Have you tried using the `GLViewImpl::EVENT_WINDOW_CLOSE` event?

**Reply by NgVThangBz** [reply]
yes i just tried it, and it works fine, thank you.

**Comment by NgVThangBz** [comment]
So, is it possible to block the application close event?

**Reply by rh101** [reply, marked as answer]
If you search the code for the event, you'll see that it is sent out from here:

```
void GLViewImpl::onGLFWWindowCloseCallback(GLFWwindow* window)
{
    bool isClose = true;
    Director::getInstance()->getEventDispatcher()->dispatchCustomEvent(GLViewImpl::EVENT_WINDOW_CLOSE, &isClose);
    if (isClose == false)
    {
        glfwSetWindowShouldClose(window, 0);
    }
}
```

It passes a pointer to a `bool`, which is used to control if it should proceed with the closure of the window.  So, if you set it to `false`, it should stop the window from closing. For example:

```
_eventListener = Director::getInstance()->getEventDispatcher()->addCustomEventListener(GLViewImpl::EVENT_WINDOW_CLOSE,
    [this](EventCustom* e) {
        if (someConditionToStopClosure) 
        {
            const auto close = static_cast<bool*>(e->getUserData());
            *close = false; // Don't allow the window to close
        }
    });
```

**Reply by NgVThangBz** [reply]
it works fine, thank you.

**Reply by rh101** [reply]
Just a side note, and if I understand your issue correctly, I feel that this is not the right solution.  If the application is crashing on exit, then you need to find out exactly why, and address the issue accordingly.

If you are getting crashes on exit because of the order of items being deleted, then that is what you need to focus on.  Fixing the issue by waiting for the window to close is just a band-aid solution, if that.  It may be a sign of a bigger issue in the code.

**Reply by NgVThangBz** [reply]
![Screenshot 2024-12-24 at 14 50 03](https://github.com/user-attachments/assets/48ae3d63-79a2-4af1-83c7-2a17eca9f97a)
when i close the application on pc. it will cause an error if the custom obj still exists,
when my custom obj is released by the system, it will release the other variables (text, sprite, button...).
however this variable was destroyed before, which leads to a crash,
i released my custom obj before the (text, sprite, button...) were released by the system
![Screenshot 2024-12-24 at 14 54 24](https://github.com/user-attachments/assets/fa0befe6-39c9-4ddc-990d-1721f3600a73)

**Reply by rh101** [reply]
That's actually a straightforward issue to fix, because you know exactly what is causing it.

If your object holds references to Axmol nodes (or any engine `ax::Object` instance), then it *must* call `retain()`, and also call `release()` when it is no longer required.  That reference must also to `nullptr` once it's released, so you can check for it (the macro `AX_SAFE_RELEASE_NULL` does both).

Your custom object should be deleting (or releasing) all references in its destructor.  If this is done, then you should not be having any issues at all.  You haven't shared any code, so I'm only guessing what would be causing this, but it's always something along these lines.

**Reply by NgVThangBz** [reply]
when closing the application, axmol variables are destroyed, so using anything related to this memory area can lead to a crash (`AX_SAFE_RELEASE_NULL` doesn't seem to help), so I try to free my custom obj before the system variables are freed

**Reply by rh101** [reply]
Try using conditional breakpoints on the reference count of the crashing node, and follow it from there.
